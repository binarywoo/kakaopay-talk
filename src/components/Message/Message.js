import React from 'react'
import { Alert, Divider, Avatar } from 'antd'
import MessageImage from '../MessageImage'
import './index.less'

const Message = ({
  isMyMessage,
  showDivider,
  showProfile,
  showTag,
  date,
  time,
  item
}) => {
  return (
    <div
      className={`message-row ${isMyMessage ? 'my-message' : 'other-message'}`}
    >
      {showDivider && <Divider className='date-divider'>{date}</Divider>}
      {!isMyMessage && showProfile && (
        <Avatar
          src={item.profileImage}
          icon={!item.profileImage ? 'user' : null}
          style={{ marginRight: 5, position: 'absolute' }}
        />
      )}
      <Alert
        className='message-box'
        message={
          item.type === 'image' ? (
            <MessageImage src={item.content} />
          ) : (
            item.content
          )
        }
      />
      {showTag && (
        <div className='message-tag'>
          <span className='fw-700'>{item.userId} Â· </span>
          {time}
        </div>
      )}
    </div>
  )
}

export default Message
