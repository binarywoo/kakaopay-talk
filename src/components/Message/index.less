.message-row {
  padding-bottom: 6px;
  &.my-message {
    text-align: right;
    .message-box {
      background-color: #ffdc00;
    }
  }
  &.other-message {
    text-align: left;
    .message-box {
      margin-left: 40px;
      background-color: #ffffff;
    }
    .message-tag {
      margin-left: 40px;
    }
  }
  .message-box {
    max-width: 80%;
    text-align: left;
    border: none;
    display: inline-block;
  }
  .message-tag {
    color: #ffffff;
    font-size: 12px;
  }
  .date-divider {
    color: #fff;
    font-size: 12px;
  }
}
