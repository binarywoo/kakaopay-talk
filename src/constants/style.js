export const COLOR = {
  PRIMARY: '#2d1c05',
  YELLOW: '#ffdc00'
}
