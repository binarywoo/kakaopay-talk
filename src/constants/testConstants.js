export const DUMMY_IMG_URL =
  'https://s.pstatic.net/static/www/img/2018/sp_search.svg'

export const DUMMY_CHAT = {
  title: 'test chat',
  lastUpdate: '2019-04-06T18:32:11.433Z',
  lastMessage: 'test message',
  key: 'testKey'
}

export const DUMMY_MESSAGE = {
  content: '하이',
  key: '-LbnOJfNoldq3SD0xRIR',
  lastUpdate: '2019-04-06T16:04:55.757Z',
  type: 'text',
  user: '-LbmTnQw-OjNS87fDbBt',
  userId: '이웅희'
}

export const DUMMY_IMAGE_MESSAGE = {
  content: '하이',
  key: '-LbnOJfNolds3SD0xRIR',
  lastUpdate: '2019-04-06T16:06:55.757Z',
  type: 'image',
  user: '-LbmTnQw-OjNS87fDbBt',
  userId: '이웅희'
}

export const DUMMY_USER = {
  key: '-LbmTnQw-OjNS87fDbBt',
  userId: '이웅희'
}
