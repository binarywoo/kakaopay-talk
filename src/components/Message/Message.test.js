import React from 'react'
import { shallow } from 'enzyme'
import Message from './index'
import { Divider, Avatar } from 'antd'
import { DUMMY_IMG_URL } from '../../constants/testConstants'
import { fromJS } from 'immutable'

describe('Message', () => {
  let wrapper = null
  const dummyProps = {
    isMyMessage: false,
    showDivider: true,
    showTag: true,
    showProfile: true,
    date: '2019-01-01',
    time: '10:10',
    item: {
      profileImage: null,
      content: 'test',
      userId: 'test'
    }
  }

  it('정상적으로 렌더링이 된다.', () => {
    wrapper = shallow(<Message {...dummyProps} />)
  })

  it('스냅샷과 매칭이 된다.', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('divider를 정상적으로 표현', () => {
    it('showDivider가 true일 경우 divider를 렌더링한다.', () => {
      expect(wrapper.find(Divider).length).toBe(1)
    })

    it('Divider 내부에 date가 정상적으료 표현된다.', () => {
      expect(
        wrapper
          .find(Divider)
          .children()
          .text()
      ).toBe('2019-01-01')
    })

    it('showDivider가 false일 경우 divider를 렌더링하지 않는다.', () => {
      wrapper = shallow(
        <Message {...Object.assign({}, dummyProps, { showDivider: false })} />
      )
      expect(wrapper.find(Divider).length).toBe(0)
    })
  })

  describe('타인의 메시지를 정상적으로 표현', () => {
    it('프로필 이미지를 보여준다.', () => {
      expect(wrapper.find(Avatar).length).toBe(1)
    })

    it('profileImage가 null일 경우 이미지 대신 아이콘을 보여준다.', () => {
      expect(wrapper.find(Avatar).prop('src')).toBeNull()
      expect(wrapper.find(Avatar).prop('icon')).toBe('user')
    })

    it('profileImage가 null이 아닐 경우 아이콘 대신 이미지를 보여준다.', () => {
      wrapper = shallow(
        <Message
          {...fromJS(dummyProps)
            .setIn(['item', 'profileImage'], DUMMY_IMG_URL)
            .toJS()}
        />
      )
      expect(wrapper.find(Avatar).prop('src')).toBe(DUMMY_IMG_URL)
      expect(wrapper.find(Avatar).prop('icon')).toBeNull()
    })
  })

  it('showTag 가 false일 때 message tag를 보여주지 않는다.', () => {
    wrapper = shallow(
      <Message {...Object.assign({}, dummyProps, { showTag: false })} />
    )
    expect(wrapper.find('.message-tag').length).toBe(0)
  })

  describe('나의 메시지를 정상적으로 표현', () => {
    it('나의 메시지일 때 프로필사진을 보여주지 않는다.', () => {
      wrapper = shallow(
        <Message {...Object.assign({}, dummyProps, { isMyMessage: true })} />
      )
      expect(wrapper.find(Avatar).length).toBe(0)
    })

    it('나의 메시지일 때에는 우측에 보여준다.', () => {
      wrapper = shallow(
        <Message {...Object.assign({}, dummyProps, { isMyMessage: true })} />
      )
      expect(wrapper.find('.message-row').hasClass('my-message')).toBe(true)
    })
  })
})
