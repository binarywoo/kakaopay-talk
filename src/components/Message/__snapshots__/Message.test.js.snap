// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message 스냅샷과 매칭이 된다. 1`] = `
<div
  className="message-row other-message"
>
  <Divider
    className="date-divider"
  >
    2019-01-01
  </Divider>
  <Avatar
    icon="user"
    shape="circle"
    size="default"
    src={null}
    style={
      Object {
        "marginRight": 5,
        "position": "absolute",
      }
    }
  />
  <Alert
    className="message-box"
    message="test"
  />
  <div
    className="message-tag"
  >
    <span
      className="fw-700"
    >
      test
       · 
    </span>
    10:10
  </div>
</div>
`;
